lib/rust/data-structures/src/counter.rs

```diff
@@ -0,0 +1,134 @@
+//! # Counter
+//!
+//! Counter type. Uses an internal global value to ensure every instance created has a different
+//! value.
+
+use crate::prelude::*;
+
+
+
+// ===============
+// === Counter ===
+// ===============
+
+/// Implements a globally-unique counter.
+#[derive(Clone, Copy, Debug, Eq, Hash, PartialEq, Ord, PartialOrd)]
+pub struct Counter {
+    value: std::num::NonZeroU64,
+}
+
+impl Counter {
+    /// Generate a unique value.
+    #[allow(clippy::new_without_default)] // Every new instance must have a different value.
+    #[allow(unsafe_code)] // See comment inside.
+    pub fn new() -> Self {
+        use std::sync::atomic;
+        static NEXT: atomic::AtomicU64 = atomic::AtomicU64::new(1);
+        let value = NEXT.fetch_add(1, atomic::Ordering::Relaxed);
```

Comment:
Do you know if the AtomicU64 WASM single-threaded environment has no overhead when compared to thread_local variables? I used the latter so far, because I feared all the syncing stuff, which is not needed in WASM.

Reply:
it compiles to plain non-atomic operations, so it shouldn't have any overhead.